File: Data Structures & Algorithms/generate-parentheses/submission-1.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        result = []

        def helper(curr, num_open):
            
            if len(curr) == 2 * n:
                result.append(curr[:])
                return
            
            num_close = len(curr) - num_open
            if num_close > num_open:
                return
            elif num_close == num_open:
                helper(curr + '(', num_open + 1)
            else:
                if num_open < n:
                    helper(curr + '(', num_open + 1)
                helper(curr + ')', num_open)
        
        helper('(', 1)

        return result

File: Data Structures & Algorithms/merge-intervals/submission-1.java
class Solution {
    public int[][] merge(int[][] intervals) {
        // sort intervals array by start position
        Arrays.sort(intervals, (a, b) -> Integer.compare(a[0], b[0]));

        ArrayList<int[]> result = new ArrayList<>();
        int i = 0, n = intervals.length;
        int low = 0, high = 0;
        boolean start = true;

        while (i < n) {
            int[] curr = intervals[i];
            if (start) {
                low = curr[0];
                high = curr[1];
                start = false;
            } else {
                if (low <= curr[0] && curr[0] <= high) {
                    high = Math.max(high, curr[1]);
                } else {
                    result.add(new int[] {low, high});
                    start = true;
                    continue;
                }
            }
            i++;
        }
        result.add(new int[] {low, high});
        return result.toArray(new int[result.size()][]);
    }
}
